File: src/python/otsim/msgbus/pusher.py
from __future__ import annotations

import json, zmq

from otsim.msgbus.envelope import Envelope

class Pusher:
  def __init__(self: Pusher, endpoint: str):
    self.ctx    = zmq.Context()
    self.socket = self.ctx.socket(zmq.PUSH)

    self.socket.connect(endpoint)
    self.socket.setsockopt(zmq.LINGER, 0)

  def push(self: Pusher, topic: str, env: Envelope) -> None:
    self.socket.send_multipart((topic.encode(), json.dumps(env).encode()))

File: src/c++/msgbus/pusher.cpp
#include "pusher.hpp"

namespace otsim {
namespace msgbus {

Pusher::Pusher(const std::string& endpoint) {
  socket = zmq::socket_t(ctx, ZMQ_PUSH);

  socket.connect(endpoint);
  socket.set(zmq::sockopt::linger, 0);
}

Pusher::~Pusher() {
  socket.close();
  ctx.close();
}

} // namespace msgbus
} // namespace otsim